default.nix

```diff
@@ -17,52 +17,45 @@ rec {
     , # Ignore the following matches
       exclude ? [ ]
     }:
-    let
-      rootStr = toString root;
+      assert builtins.isPath root;
+      let
+        rootStr = toString root;
 
-      # If an argument to include or exclude is a path, transform it to a matcher.
-      #
-      # This probably needs more work, I don't think that it works on
-      # sub-folders.
-      toMatcher = f:
-        let
-          # Push these here to memoize the result
-          path_ = toString f;
-          path__ = "${rootStr}/${f}";
-        in
-        if builtins.isFunction f then f
-        else if builtins.isPath f then (_: path: _: path_ == path)
-        else if builtins.isString f then (_: path: _: path__ == path)
-        else
-          throw "Unsupported type ${builtins.typeOf f}";
+        # If an argument to include or exclude is a path, transform it to a matcher.
+        #
+        # This probably needs more work, I don't think that it works on
+        # sub-folders.
+        toMatcher = f:
+          let
+            path_ = _toCleanPath root f;
+          in
+          if builtins.isFunction f then f root
+          else
+            (path: _: path_ == path);
 
-      include_ = map toMatcher include;
-      exclude_ = map toMatcher exclude;
-    in
-    builtins.path {
-      inherit name;
-      path = root;
-      filter = path: type:
-        (builtins.any (f: f root path type) include_) &&
-        (!builtins.any (f: f root path type) exclude_);
-    };
+        include_ = map toMatcher include;
+        exclude_ = map toMatcher exclude;
+      in
+      builtins.path {
+        inherit name;
+        path = root;
+        filter = path: type:
+          (builtins.any (f: f path type) include_) &&
+          (!builtins.any (f: f path type) exclude_);
+      };
 
   # Match a directory and any path inside of it
-  inDirectory = directory:
-    root: path: type:
-      let
-        # Convert `directory` to a path to clean user input
-        dirAsPath =
-          if builtins.isString directory then root + ("/" + directory)
-          else if builtins.isPath directory then directory
-          else
-            throw "inDirectory: Unsupported type ${builtins.typeOf directory}, expected string or path";
-        directory_ = toString dirAsPath;
-        path_ = toString path;
-      in
-      directory_ == path_
-      # Add / to the end to make sure we match a full directory prefix
-      || _hasPrefix (directory_ + "/") path_;
+  inDirectory =
+    directory:
+    root:
+    let
+      # Convert `directory` to a path to clean user input.
+      directory_ = _toCleanPath root directory;
+    in
+    path: type:
+    directory_ == path
```

Comment:
This will always fail since `directory_` is a string and `path` is a path. Need a `toString` on `path`.

Reply:
I thought the same as you until I started writing this library. It turns out that the path argument is already a string!

```
nix-repl> builtins.path { path = ./.; filter = path: _: builtins.trace "path=${path} type=${builtins.typeOf path}" false; }  
trace: path=/home/zimbatm/go/src/github.com/numtide/nix-filter/tests/default.nix type=string
trace: path=/home/zimbatm/go/src/github.com/numtide/nix-filter/tests/fixture1 type=string
```

Reply:
Huh... good to know!